extensions/cornerstone/src/commandsModule.ts

```diff
@@ -423,6 +378,225 @@ const commandsModule = ({ servicesManager }) => {
+      if (!cachedLayout || !cachedLayout[cacheId]) {
+        return;
+      }
+
+      const { viewports: cachedViewports, numRows, numCols } = cachedLayout[
+        cacheId
+      ];
+
+      // Todo: The following assumes that when turning off MPR we are applying the default
+      //  protocol which might not be the one that was used before MPR was turned on
+      // In order to properly implement this logic, we should modify the hanging protocol
+      // upon layout change with layout selector, and cache and restore it when turning
+      // MPR on and off
+      const viewportStructure = getProtocolViewportStructureFromGridViewports({
+        viewports: cachedViewports,
+        numRows,
+        numCols,
+      });
+
+      const viewportSpecificMatch = cachedViewports.reduce(
+        (acc, viewport, index) => {
+          const {
+            displaySetInstanceUIDs,
+            viewportOptions,
+            displaySetOptions,
+          } = viewport;
+
+          acc[index] = {
+            displaySetInstanceUIDs,
+            viewportOptions,
+            displaySetOptions,
+          };
+
+          return acc;
+        },
+        {}
+      );
+
+      const defaultProtocol = HangingProtocolService.getProtocolById('default');
+
+      // Todo: this assumes there is only one stage in the default protocol
+      const defaultProtocolStage = defaultProtocol.stages[0];
+      defaultProtocolStage.viewportStructure = viewportStructure;
+
+      const { primaryToolId } = ToolBarService.state;
+      const mprToolGroup = _getToolGroup(MPR_TOOLGROUP_ID);
+      // turn off crosshairs if it is on
+      if (
+        primaryToolId === 'Crosshairs' ||
+        mprToolGroup.getToolInstance('Crosshairs')?.mode ===
+          Enums.ToolModes.Active
+      ) {
+        const toolGroup = _getToolGroup(MPR_TOOLGROUP_ID);
+        toolGroup.setToolDisabled('Crosshairs');
+        ToolBarService.recordInteraction({
+          groupId: 'WindowLevel',
+          itemId: 'WindowLevel',
+          interactionType: 'tool',
+          commands: [
+            {
+              commandName: 'setToolActive',
+              commandOptions: {
+                toolName: 'WindowLevel',
+              },
+              context: 'CORNERSTONE',
+            },
+          ],
+        });
+      }
+
+      // clear segmentations if they exist
+      removeToolGroupSegmentationRepresentations(MPR_TOOLGROUP_ID);
+
+      HangingProtocolService.setProtocol(
+        'default',
+        viewportSpecificMatch,
+        error => {
+          UINotificationService.show({
+            title: 'Multiplanar reconstruction (MPR) ',
+            message:
+              'Something went wrong while trying to restore the previous layout.',
+            type: 'info',
+            duration: 3000,
+          });
+        }
+      );
+    },
+    toggleStackImageSync: ({ toggledState }) => {
+      if (!toggledState) {
+        STACK_IMAGE_SYNC_GROUPS_INFO.forEach(syncGroupInfo => {
```

Comment:
Commands module is getting pretty big, maybe it should be split into multiple files. Kind of ugly to start putting global variables at the top of the file like STACK_IMAGE_SYNC_GROUPS_INFO

Reply:
Yeah, I don't like the STACK_IMAGE_SYNC_GROUPS_INFO either, but this is a kind of a global thing that we need, where else do you think I should put it?